src/nvim/api/vim.c

```diff
@@ -2851,3 +2852,117 @@ void nvim_set_decoration_provider(Integer ns_id, DictionaryOf(LuaRef) opts,
 error:
   clear_provider(p);
 }
+
+///
+/// Set matches and start insert completion: see |complete()|.
+///
+/// @param startcol: Byte offset where the completed text starts.
+///                     Any text between startcol and the current
+///                     cursor col is replaced by the completion.
+/// @param matches: A list of string matches.
+/// @param opts: Dictionary. Possible keys:
+///                 filterfunc: function used for filtering the
+///                     matches. See |nvim_register_filterfunc|.
+//
+// todo: completefunc with refresh: always
+// todo: figure out what to do with ctrl_e and ctrl_l
+// todo: what to do if filterfunc != NULL AND completion entry
+// has icase/equal set.
+String nvim_complete(Integer startcol,
+                   Object matches,
+                   DictionaryOf(LuaRef) opts,
+                   Error *err)
+  FUNC_API_SINCE(7)
+{
+  typval_T tv;
+  extern LuaRef *user_filterfunc;
+  extern LuaRef *active_filterfunc;
+  LuaRef filterfunc;
+
+  if (matches.type != kObjectTypeArray) {
+    api_set_error(err, kErrorTypeValidation,
+              "wrong type");
+  }
+
+  if ((State & INSERT) == 0) {
+    api_set_error(err, kErrorTypeValidation,
+            "complete() can only be used in Insert mode");
+    goto error;
+  }
+
+  // Check for undo allowed here, because if something was already inserted
+  // the line was already saved for undo and this check isn't done.
+  if (!undo_allowed())
+      goto error;
+
+  if (startcol <= 0) {
+      goto error;
+  }
+
+  for (size_t i = 0; i < opts.size; i++) {
+    String k = opts.items[i].key;
+    Object *v = &opts.items[i].value;
+
+    if (strequal(k.data, "filterfunc")){
+      if (v->type != kObjectTypeLuaRef) {
+        api_set_error(err, kErrorTypeValidation,
+                "expected lua function");
+        goto error;
+      }
+      filterfunc = api_new_luaref(v->data.luaref);
```

Comment:
the canonical pattern is `filterfunc = v->data.luaref; v->data.luaref = LUA_NOREF`